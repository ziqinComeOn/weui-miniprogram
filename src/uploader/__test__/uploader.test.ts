import path from 'path'
import simulate from 'miniprogram-simulate'
;(globalThis as any).wx = (globalThis as any).wx || {}
;(globalThis as any).wx.chooseImage = jest.fn(({ success }) => {
    setTimeout(() => {
        success &&
            success({ tempFilePaths: ['/a/b.png'], tempFiles: [{ path: '/a/b.png', size: 1024 }] })
    }, 0)
})
;(globalThis as any).wx.getFileSystemManager = jest.fn(() => ({
    readFileSync: jest.fn(() => new ArrayBuffer(1024))
}))
;(globalThis as any).wx.arrayBufferToBase64 = jest.fn((buffer: ArrayBuffer) => {
    let binary = ''
    const bytes = new Uint8Array(buffer)
    const len = bytes.byteLength
    for (let i = 0; i < len; i++) {
        binary += String.fromCharCode(bytes[i])
    }
    return window.btoa(binary)
})

describe('uploader', () => {
    const gallery = simulate.load({ template: '<slot /> ' })

    const uploader = simulate.load(path.resolve(__dirname, '../uploader'), 'mp-uploader', {
        usingComponents: {
            'mp-gallery': gallery
        }
    })

    test('basic', async () => {
        const comp = simulate.render(
            simulate.load({
                usingComponents: {
                    'mp-uploader': uploader,
                    'mp-gallery': gallery
                },
                template: `
                    <mp-uploader id="uploader" upload="{{uploadFile}}" files="{{files}}" max-count="5" title="图片上传" tips="图片上传提示"></mp-uploader>
                `,
                data: {
                    files: []
                },
                ready() {
                    this.setData({
                        uploadFile: this.uploadFile.bind(this)
                    })
                },
                methods: {
                    uploadFile(files) {
                        return Promise.resolve({ urls: ['http://qq.com/b.png'] })
                    }
                }
            })
        )
        comp.attach(document.createElement('parent-wrapper'))
        await simulate.sleep(0)
        expect(comp.toJSON()).toMatchSnapshot()

        const uploaderComp = comp.querySelector('#uploader')
        const chooseImg = uploaderComp.querySelector('.weui-uploader__input')
        chooseImg.dispatchEvent('tap')
        await simulate.sleep(100)
        expect(comp.toJSON()).toMatchSnapshot()
    })

    test('upload fail', async () => {
        const comp = simulate.render(
            simulate.load({
                usingComponents: {
                    'mp-uploader': uploader,
                    'mp-gallery': gallery
                },
                template: `
                    <mp-uploader id="uploader" upload="{{uploadFile}}" files="{{files}}" max-count="5" title="图片上传" tips="图片上传提示"></mp-uploader>
                `,
                data: {
                    files: []
                },
                ready() {
                    this.setData({
                        uploadFile: this.uploadFile.bind(this)
                    })
                },
                methods: {
                    uploadFile(files) {
                        return Promise.reject('error')
                    }
                }
            })
        )
        comp.attach(document.createElement('parent-wrapper'))
        await simulate.sleep(0)
        expect(comp.toJSON()).toMatchSnapshot()

        const uploaderComp = comp.querySelector('#uploader')
        const chooseImg = uploaderComp.querySelector('.weui-uploader__input')
        chooseImg.dispatchEvent('tap')
        await simulate.sleep(100)
        expect(comp.toJSON()).toMatchSnapshot()
    })

    test('preview', async () => {
        const comp = simulate.render(uploader, {
            files: [
                {
                    url: 'http://qq.com/image.png'
                }
            ],
            maxCount: '5',
            title: '图片上传',
            tips: '图片上传提示'
        })
        comp.attach(document.createElement('parent-wrapper'))
        await simulate.sleep(0)
        expect(comp.toJSON()).toMatchSnapshot()

        const files = comp.querySelectorAll('.weui-uploader__file')
        files[0].dispatchEvent('tap')
        await simulate.sleep(0)
        expect(comp.toJSON()).toMatchSnapshot()

        const gallery = comp.querySelector('.gallery')
        gallery.dispatchEvent('delete', { url: 'http://qq.com/image.png', index: 0 })
        await simulate.sleep(0)
        expect(comp.toJSON()).toMatchSnapshot()
    })
})
