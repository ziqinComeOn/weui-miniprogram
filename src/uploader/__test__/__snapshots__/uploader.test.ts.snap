// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uploader basic 1`] = `
<main>
  <mp-uploader
    id="uploader"
  >
    <wx-view
      class="weui-uploader "
    >
      <wx-view
        class="weui-uploader__hd"
      >
        <wx-view
          class="weui-uploader__overview"
        >
          <wx-view
            class="weui-uploader__title"
          >
            图片上传
          </wx-view>
          <wx-view
            class="weui-uploader__info"
          >
            0/5
          </wx-view>
        </wx-view>
        <wx-view
          class="weui-uploader__tips"
        >
          图片上传提示
        </wx-view>
      </wx-view>
      <wx-view
        class="weui-uploader__bd"
      >
        <wx-view
          class="weui-uploader__files"
        />
        <wx-view
          class="weui-uploader__input-box"
        >
          <wx-view
            class="weui-uploader__input"
            bind:tap="chooseImage"
          />
        </wx-view>
      </wx-view>
    </wx-view>
    <mp-gallery
      class="gallery"
      current=""
      hideOnClick="{{true}}"
      imgUrls="{{Array []}}"
      show="{{false}}"
      showDelete="{{true}}"
      bind:delete="deletePic"
    />
  </mp-uploader>
</main>
`;

exports[`uploader basic 2`] = `
<main>
  <mp-uploader
    id="uploader"
  >
    <wx-view
      class="weui-uploader "
    >
      <wx-view
        class="weui-uploader__hd"
      >
        <wx-view
          class="weui-uploader__overview"
        >
          <wx-view
            class="weui-uploader__title"
          >
            图片上传
          </wx-view>
          <wx-view
            class="weui-uploader__info"
          >
            1/5
          </wx-view>
        </wx-view>
        <wx-view
          class="weui-uploader__tips"
        >
          图片上传提示
        </wx-view>
      </wx-view>
      <wx-view
        class="weui-uploader__bd"
      >
        <wx-view
          class="weui-uploader__files"
        >
          <wx-view
            class="weui-uploader__file weui-uploader__file_status"
            data-index="{{0}}"
            bind:tap="previewImage"
          >
            <wx-image
              class="weui-uploader__img"
              mode="aspectFill"
              src="data:image/jpg;base64,AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=="
            />
            <wx-view
              class="weui-uploader__file-content"
            >
              <wx-view
                class="weui-loading"
              />
            </wx-view>
          </wx-view>
        </wx-view>
        <wx-view
          class="weui-uploader__input-box"
        >
          <wx-view
            class="weui-uploader__input"
            bind:tap="chooseImage"
          />
        </wx-view>
      </wx-view>
    </wx-view>
    <mp-gallery
      class="gallery"
      current=""
      hideOnClick="{{true}}"
      imgUrls="{{Array []}}"
      show="{{false}}"
      showDelete="{{true}}"
      bind:delete="deletePic"
    />
  </mp-uploader>
</main>
`;

exports[`uploader preview 1`] = `
<mp-uploader>
  <wx-view
    class="weui-uploader "
  >
    <wx-view
      class="weui-uploader__hd"
    >
      <wx-view
        class="weui-uploader__overview"
      >
        <wx-view
          class="weui-uploader__title"
        >
          图片上传
        </wx-view>
        <wx-view
          class="weui-uploader__info"
        >
          1/5
        </wx-view>
      </wx-view>
      <wx-view
        class="weui-uploader__tips"
      >
        图片上传提示
      </wx-view>
    </wx-view>
    <wx-view
      class="weui-uploader__bd"
    >
      <wx-view
        class="weui-uploader__files"
      >
        <wx-view
          class="weui-uploader__file"
          data-index="{{0}}"
          bind:tap="previewImage"
        >
          <wx-image
            class="weui-uploader__img"
            mode="aspectFill"
            src="http://qq.com/image.png"
          />
        </wx-view>
      </wx-view>
      <wx-view
        class="weui-uploader__input-box"
      >
        <wx-view
          class="weui-uploader__input"
          bind:tap="chooseImage"
        />
      </wx-view>
    </wx-view>
  </wx-view>
  <mp-gallery
    class="gallery"
    current=""
    hideOnClick="{{true}}"
    imgUrls="{{Array []}}"
    show="{{false}}"
    showDelete="{{true}}"
    bind:delete="deletePic"
  />
</mp-uploader>
`;

exports[`uploader preview 2`] = `
<mp-uploader>
  <wx-view
    class="weui-uploader "
  >
    <wx-view
      class="weui-uploader__hd"
    >
      <wx-view
        class="weui-uploader__overview"
      >
        <wx-view
          class="weui-uploader__title"
        >
          图片上传
        </wx-view>
        <wx-view
          class="weui-uploader__info"
        >
          1/5
        </wx-view>
      </wx-view>
      <wx-view
        class="weui-uploader__tips"
      >
        图片上传提示
      </wx-view>
    </wx-view>
    <wx-view
      class="weui-uploader__bd"
    >
      <wx-view
        class="weui-uploader__files"
      >
        <wx-view
          class="weui-uploader__file"
          data-index="{{0}}"
          bind:tap="previewImage"
        >
          <wx-image
            class="weui-uploader__img"
            mode="aspectFill"
            src="http://qq.com/image.png"
          />
        </wx-view>
      </wx-view>
      <wx-view
        class="weui-uploader__input-box"
      >
        <wx-view
          class="weui-uploader__input"
          bind:tap="chooseImage"
        />
      </wx-view>
    </wx-view>
  </wx-view>
  <mp-gallery
    class="gallery"
    current="{{0}}"
    hideOnClick="{{true}}"
    imgUrls="{{
      Array [
        "http://qq.com/image.png",
      ]
    }}"
    show="{{true}}"
    showDelete="{{true}}"
    bind:delete="deletePic"
  />
</mp-uploader>
`;

exports[`uploader preview 3`] = `
<mp-uploader>
  <wx-view
    class="weui-uploader "
  >
    <wx-view
      class="weui-uploader__hd"
    >
      <wx-view
        class="weui-uploader__overview"
      >
        <wx-view
          class="weui-uploader__title"
        >
          图片上传
        </wx-view>
        <wx-view
          class="weui-uploader__info"
        >
          0/5
        </wx-view>
      </wx-view>
      <wx-view
        class="weui-uploader__tips"
      >
        图片上传提示
      </wx-view>
    </wx-view>
    <wx-view
      class="weui-uploader__bd"
    >
      <wx-view
        class="weui-uploader__files"
      />
      <wx-view
        class="weui-uploader__input-box"
      >
        <wx-view
          class="weui-uploader__input"
          bind:tap="chooseImage"
        />
      </wx-view>
    </wx-view>
  </wx-view>
  <mp-gallery
    class="gallery"
    current="{{0}}"
    hideOnClick="{{true}}"
    imgUrls="{{
      Array [
        "http://qq.com/image.png",
      ]
    }}"
    show="{{true}}"
    showDelete="{{true}}"
    bind:delete="deletePic"
  />
</mp-uploader>
`;

exports[`uploader upload fail 1`] = `
<main>
  <mp-uploader
    id="uploader"
  >
    <wx-view
      class="weui-uploader "
    >
      <wx-view
        class="weui-uploader__hd"
      >
        <wx-view
          class="weui-uploader__overview"
        >
          <wx-view
            class="weui-uploader__title"
          >
            图片上传
          </wx-view>
          <wx-view
            class="weui-uploader__info"
          >
            0/5
          </wx-view>
        </wx-view>
        <wx-view
          class="weui-uploader__tips"
        >
          图片上传提示
        </wx-view>
      </wx-view>
      <wx-view
        class="weui-uploader__bd"
      >
        <wx-view
          class="weui-uploader__files"
        />
        <wx-view
          class="weui-uploader__input-box"
        >
          <wx-view
            class="weui-uploader__input"
            bind:tap="chooseImage"
          />
        </wx-view>
      </wx-view>
    </wx-view>
    <mp-gallery
      class="gallery"
      current=""
      hideOnClick="{{true}}"
      imgUrls="{{Array []}}"
      show="{{false}}"
      showDelete="{{true}}"
      bind:delete="deletePic"
    />
  </mp-uploader>
</main>
`;

exports[`uploader upload fail 2`] = `
<main>
  <mp-uploader
    id="uploader"
  >
    <wx-view
      class="weui-uploader "
    >
      <wx-view
        class="weui-uploader__hd"
      >
        <wx-view
          class="weui-uploader__overview"
        >
          <wx-view
            class="weui-uploader__title"
          >
            图片上传
          </wx-view>
          <wx-view
            class="weui-uploader__info"
          >
            1/5
          </wx-view>
        </wx-view>
        <wx-view
          class="weui-uploader__tips"
        >
          图片上传提示
        </wx-view>
      </wx-view>
      <wx-view
        class="weui-uploader__bd"
      >
        <wx-view
          class="weui-uploader__files"
        >
          <wx-view
            class="weui-uploader__file weui-uploader__file_status"
            data-index="{{0}}"
            bind:tap="previewImage"
          >
            <wx-image
              class="weui-uploader__img"
              mode="aspectFill"
              src="data:image/jpg;base64,AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=="
            />
            <wx-view
              class="weui-uploader__file-content"
            >
              <wx-view
                class="weui-loading"
              />
            </wx-view>
          </wx-view>
        </wx-view>
        <wx-view
          class="weui-uploader__input-box"
        >
          <wx-view
            class="weui-uploader__input"
            bind:tap="chooseImage"
          />
        </wx-view>
      </wx-view>
    </wx-view>
    <mp-gallery
      class="gallery"
      current=""
      hideOnClick="{{true}}"
      imgUrls="{{Array []}}"
      show="{{false}}"
      showDelete="{{true}}"
      bind:delete="deletePic"
    />
  </mp-uploader>
</main>
`;
